In [72]:
%reload_ext autoreload
%autoreload 2

In [73]:
%reload_ext autoreload

In [74]:
import pandas as pd
import numpy as np  
import matplotlib.pyplot as plt 
import os
import sys
import json

In [69]:
os.getcwd()

'/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis'

In [5]:
# Add src to system path to import data_loader
sys.path.append(os.path.abspath(os.path.join('..', 'src')))
sys.path.append(os.path.abspath(os.path.join('..')))

In [6]:
sys.path

['/usr/lib/python38.zip',
 '/usr/lib/python3.8',
 '/usr/lib/python3.8/lib-dynload',
 '',
 '/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis/adluido/lib/python3.8/site-packages',
 '/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis/src',
 '/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis']

In [7]:
os.chdir("..")
os.getcwd()

'/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis'

In [8]:
filepath = "data/adluido/Challenge_Data/performance_data.csv"
assets_dir = '/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis/data/adluido/Challenge_Data/Assets'
categories_path = 'data/adluido/categories.txt'
concepts_path = 'data/adluido/concepts.json'
performance_data_path = 'data/adluido/Challenge_Data/performance_data.csv'

In [7]:
from src.feature_engineering import extract_features, apply_pca, load_image_batches

In [8]:
# Load processed image data
image_data = load_image_batches('.')
print(f"Loaded {len(image_data)} processed images")

# Extract features
print("Extracting features...")
features = extract_features(image_data)
print(f"Extracted features shape: {features.shape}")

# Apply PCA
print("Applying PCA...")
pca_features = apply_pca(features)
print(f"PCA features shape: {pca_features.shape}")

# Save features
np.save('image_features.npy', pca_features)
print("Saved image features to image_features.npy")

Loaded 10000 processed images
Extracting features...


Extracting features: 100%|██████████| 10000/10000 [05:25<00:00, 30.74it/s]


Extracted features shape: (10000, 34336)
Applying PCA...
PCA features shape: (10000, 100)
Saved image features to image_features.npy


### Feature Analysis and Visualization

In [15]:
from src.feature_analysis import load_features, visualize_pca_variance, apply_tsne, visualize_tsne, apply_kmeans, visualize_kmeans

In [17]:
# Load the features
features = load_features('image_features.npy')
print(f"Loaded features with shape: {features.shape}")

Loaded features with shape: (10000, 100)


In [18]:
# Visualize PCA variance
visualize_pca_variance(features)
print("PCA variance visualization saved as 'pca_variance.png'")

PCA variance visualization saved as 'pca_variance.png'


Cumulative Explained Variance Ratio vs. Number of PCA Components:
This graph shows how much of the total variance in our data is explained by each principal component. We can see that:


The curve rises steeply at first and then starts to level off
Around 20-30 components explain about 80% of the variance
Using 100 components (as we did) captures over 95% of the variance

This suggests that our PCA reduction to 100 components was effective in preserving most of the important information while significantly reducing dimensionality.

![alt text](../pca_variance.png)

In [19]:
# Apply t-SNE
tsne_results = apply_tsne(features)
visualize_tsne(tsne_results)
print("t-SNE visualization saved as 'tsne_visualization.png'")

t-SNE visualization saved as 'tsne_visualization.png'


t-SNE visualization of image features
This plot shows the distribution of our image features in a 2D space. We can observe:

A complex, non-linear structure in the data
Some clear clusters and subgroups
A generally even spread of points, with some denser regions

This visualization confirms that our images have diverse features and that there are underlying patterns in the data.

![alt text](../tsne_visualization.png)

In [20]:
# Apply K-means clustering
cluster_labels = apply_kmeans(features)
visualize_kmeans(tsne_results, cluster_labels)
print("K-means clustering visualization saved as 'kmeans_clustering.png'")

print("Analysis complete. Check the generated PNG files for visualizations.")

/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis/adluido/lib/python3.8/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


K-means clustering visualization saved as 'kmeans_clustering.png'
Analysis complete. Check the generated PNG files for visualizations.


K-means clustering of image features (t-SNE visualization):
This plot shows distinct clusters of image features, indicating that our feature extraction process has successfully captured meaningful differences between images. We can observe that


5 main clusters, each represented by a different color
Some overlap between clusters, suggesting similarities between certain groups of images
A spread of data points, indicating a diverse range of features

This clustering could represent different types of images or elements within our ad creatives (e.g., text-heavy images, product shots, people, backgrounds, etc.).

![alt text](../kmeans_clustering.png)

Based on these visualizations, we can conclude that

There are distinct groups or types of images in our dataset.
The PCA reduction preserved most of the important information.
The data has a complex structure that could be challenging but informative for our Critic/Grading Agent.

In [17]:
from src.advanced_image_analysis import load_data, investigate_clusters, analyze_feature_importance, correlate_with_performance

### Advanced Analysis

In [10]:
image_features, performance_data = load_data()

# Investigate clusters
cluster_labels, closest_points = investigate_clusters(image_features)
print(f"Cluster sizes: {np.bincount(cluster_labels)}")
print(f"Indices of closest points to cluster centers: {closest_points}")

/home/abubeker_shamil/Automated_Ad_Storyboard_Synthesis/adluido/lib/python3.8/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


Cluster sizes: [1001 2012 3434 1233 2320]
Indices of closest points to cluster centers: [280, 113, 2117, 692, 619]


#### Cluster Sizes
Cluster 0: 1001
Cluster 1: 2012
Cluster 2: 3434
Cluster 3: 1233
Cluster 4: 2320

This shows that Cluster 2 is the largest, containing about 34% of the images, while Cluster 0 is the smallest with about 10% of the images.
#### Indices of closest points to cluster centers: [280, 113, 2117, 692, 619]
These indices represent the images that are most representative of each cluster. We could use these as exemplars when visually analyzing the characteristics of each cluster.
#### Average ER and CTR by Cluster
Cluster 0: ER = 0.232770, CTR = 0.107503 (Lowest ER, Second-lowest CTR)
Cluster 1: ER = 0.276470, CTR = 0.116850 (Highest ER, Third-highest CTR)
Cluster 2: ER = 0.261782, CTR = 0.112727 (Middle ER, Lowest CTR)
Cluster 3: ER = 0.258947, CTR = 0.131149 (Second-lowest ER, Highest CTR)
Cluster 4: ER = 0.263384, CTR = 0.116623 (Middle ER, Second-highest CTR)
#### Sample feature values for each cluster
This data shows the values of the first 7 features for the central point of each cluster. We can see significant differences between clusters, which explains why they were separated.

In [11]:
# Analyze feature importance
n_components_90 = analyze_feature_importance(image_features)
print(f"Number of components needed to explain 90% of variance: {n_components_90}")

Number of components needed to explain 90% of variance: 51


PCA: Cumulative Explained Variance Ratio
The graph shows that 90% of the variance in our data can be explained by 51 components. This is a significant reduction from our original feature space while still retaining most of the information. It suggests that we can effectively reduce the dimensionality of our data from 100 to 51 components without losing much important information.

![alt text](../pca_explained_variance.png)

In [40]:
# Correlate with performance
cluster_performance, merged_data = correlate_with_performance(image_features, performance_data, cluster_labels)
print("Average ER and CTR by cluster:")

print("Analysis complete. Check the generated PNG files for visualizations.")

Number of samples with both image features and performance data: 905
Average ER and CTR by cluster:
Analysis complete. Check the generated PNG files for visualizations.


In [42]:

cluster_performance.to_csv('cluster_performance.csv')

Average ER and CTR by Cluster
This graph provides valuable insights into the performance of different image clusters


Cluster 1 has the highest Engagement Rate (ER) but not the highest Click-Through Rate (CTR).
Cluster 3 has the highest CTR, though its ER is slightly lower than Cluster 1.
Cluster 0 has the lowest performance in both ER and CTR.
Clusters 2 and 4 show moderate performance in both metrics.

![alt text](../cluster_performance.png)

Cluster Analysis:

**Cluster** 1 (20% of images) performs best overall, with the highest ER and good CTR. This should be a primary focus for our Critic/Grading Agent.

**Cluster** 3 (12% of images) has the highest CTR but lower ER. It could represent images that are very effective at generating clicks once a user engages.

**Cluster** 2 (34% of images) has middling ER but the lowest CTR. This large cluster might represent "average" images that engage users but don't drive clicks effectively.

**Cluster** 0 (10% of images) performs poorest overall and should be studied to understand what to avoid.

**Cluster** 4 (23% of images) shows balanced performance and could represent a safe, middle-ground approach.

### Cluster Analysis and Visualization

In [84]:
from src.cluster_visualization import visualize_cluster_representatives, load_image_data

In [107]:
assets_dir = 'data/adluido/Challenge_Data/Assets'
indices = [280, 113, 2117, 692, 619]  # Indices of closest points to cluster centers

image_data = load_image_data(assets_dir)
print(f"Loaded {len(image_data)} images")

# Correct cluster performance data
cluster_performance = pd.read_csv('cluster_performance.csv')

visualize_cluster_representatives(indices, image_data, cluster_performance)
print("Cluster representative images have been saved as 'cluster_representatives.png'")

Loaded 16273 images
Index From Cluster Visualization is : 20 and len(image_data) : 16273
The Preview image path is :: data/adluido/Challenge_Data/Assets/248ad697696f5e5ef82b968bfe5480ba/_preview.png image folder path is :: data/adluido/Challenge_Data/Assets/248ad697696f5e5ef82b968bfe5480ba
Index From Cluster Visualization is : 11 and len(image_data) : 16273
The Preview image path is :: data/adluido/Challenge_Data/Assets/05dff647236dcc065f51eea259e8a120/_preview.png image folder path is :: data/adluido/Challenge_Data/Assets/05dff647236dcc065f51eea259e8a120
Index From Cluster Visualization is : 50 and len(image_data) : 16273
The Preview image path is :: data/adluido/Challenge_Data/Assets/bcc5ae4665b6b8c2ff25dfdc65fbc65a/_preview.png image folder path is :: data/adluido/Challenge_Data/Assets/bcc5ae4665b6b8c2ff25dfdc65fbc65a
Index From Cluster Visualization is : 67 and len(image_data) : 16273
The Preview image path is :: data/adluido/Challenge_Data/Assets/bcc5ae4665b6b8c2ff25dfdc65fbc65a/_

## Corelation Analysis

In [103]:
from src.corelation_analysis import load_data, correlate_features_with_performance, visualize_top_features, analyze_performance_distribution

In [105]:
image_features, performance_data = load_data()

correlation_er, correlation_ctr = correlate_features_with_performance(image_features, performance_data)
visualize_top_features(correlation_er, correlation_ctr)

analyze_performance_distribution(performance_data)

print("Top 5 features correlated with ER:")
print(correlation_er[:5])
print("\nTop 5 features correlated with CTR:")
print(correlation_ctr[:5])

print("\nAnalysis complete. Check the generated PNG files for visualizations.")

Top 5 features correlated with ER:
ER     1.000000
CTR    0.453146
20     0.096262
11     0.069027
50     0.064961
Name: ER, dtype: float64

Top 5 features correlated with CTR:
CTR    1.000000
ER     0.453146
67     0.087513
20     0.084923
87     0.079520
Name: CTR, dtype: float64

Analysis complete. Check the generated PNG files for visualizations.
